dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp

```diff
@@ -0,0 +1,988 @@
+{
+    if (!SharedHelpers::IsRS5OrHigher())
+    {
+        return;
+    }
+
+    if (m_nowPlaying)
+    {
+        // Stop the animation by setting the Progress value to the fromProgress of the
+        // most recent play.
+        SetProgress(m_currentPlayFromProgress);
+    }
+}
+
+void AnimatedVisualPlayer::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& /*args*/)
+{
+    // This method is called by AnimatedVisualPlayerProperties::OnPropertyChanged which is auto generated.
+    // We have this empty function here in order to make compiler build successfully.
+    return;
+}
+
+void AnimatedVisualPlayer::OnAutoPlayPropertyChanged(
+    winrt::DependencyObject const& sender,
+    winrt::DependencyPropertyChangedEventArgs const& args)
+{
+    auto newValue = unbox_value<bool>(args.NewValue());
+
+    winrt::get_self<AnimatedVisualPlayer>(sender.as<winrt::AnimatedVisualPlayer>())->OnAutoPlayPropertyChanged(newValue);
+}
+
+
+void AnimatedVisualPlayer::OnAutoPlayPropertyChanged(bool newValue)
+{
+    if (newValue && IsAnimatedVisualLoaded() && !m_nowPlaying)
+    {
+        // Start playing immediately.
+        auto from = 0;
+        auto to = 1;
+        auto looped = true;
+        PlayAsync(from, to, looped);
+    }
+}
+
+void AnimatedVisualPlayer::OnFallbackContentPropertyChanged(
+    winrt::DependencyObject const& sender,
+    winrt::DependencyPropertyChangedEventArgs const& args)
+{
+    winrt::get_self<AnimatedVisualPlayer>(sender.as<winrt::AnimatedVisualPlayer>())->OnFallbackContentPropertyChanged();
+}
+
+
+void AnimatedVisualPlayer::OnFallbackContentPropertyChanged()
+{
+    if (m_isFallenBack)
+    {
+        LoadFallbackContent();
+    }
+}
+
+void AnimatedVisualPlayer::OnSourcePropertyChanged(
+    winrt::DependencyObject const& sender,
+    winrt::DependencyPropertyChangedEventArgs const& args)
+{
+    auto oldValue = safe_cast<winrt::IAnimatedVisualSource>(args.OldValue());
+    auto newValue = safe_cast<winrt::IAnimatedVisualSource>(args.NewValue());
+
+    winrt::get_self<AnimatedVisualPlayer>(sender.as<winrt::AnimatedVisualPlayer>())->OnSourcePropertyChanged(
+        oldValue,
+        newValue);
+}
+
+void AnimatedVisualPlayer::OnSourcePropertyChanged(winrt::IAnimatedVisualSource const& oldSource, winrt::IAnimatedVisualSource const& newSource)
+{
+    CompleteCurrentPlay();
+
+    if (auto oldDynamicSource = oldSource.try_as<winrt::IDynamicAnimatedVisualSource>())
+    {
+        // Disconnect from the update notifications of the old source.
+        oldDynamicSource.AnimatedVisualInvalidated(m_dynamicAnimatedVisualInvalidatedToken);
+        m_dynamicAnimatedVisualInvalidatedToken = { 0 };
+    }
+
+    if (auto newDynamicSource = newSource.try_as<winrt::IDynamicAnimatedVisualSource>())
+    {
+        // Connect to the update notifications of the new source.
+        m_dynamicAnimatedVisualInvalidatedToken = newDynamicSource.AnimatedVisualInvalidated([this](auto const& /*sender*/, auto const&)
```

Comment:
I don't see anything that disconnects this callback when the AnimatedVisualPlayer gets destroyed. Can you use auto_revoke instead?